### PR TITLE
Release 1.0.0

### PR DESCRIPTION
**Pull request by briandoconnor**
Typically you don't make a pull request for a release branch in [HubFlow](https://datasift.github.io/gitflow/) but I thought it would be helpful in soliciting comments.  When the release branch is ready, I'll delete this pull request after releasing in the HubFlow way.

For the time being, please leave comments below.  If you have changes to propose please make a pull request against the release/1.0.0 branch and I'll take a look.

### CODE REVIEW

**Comment by denis-yuen** [comment]
Note from meeting: 

- make metadata call return api minor version support (e.g. 1.0.0) as well as registry api

File: registry.json [hunk 1]
```diff
@@ -1,3 +1,5 @@
 {
-  "dockstore" : "https://www.dockstore.org:8443"
+  "dockstore" : [
+      "https://www.dockstore.org:8443/api/v1/metadata"
+    ]
```
Comment (tfmorris) [inline registry.json]: I don't remember all the discussion from the call. If there are multiple URLs listed here, what's the significance? What order do they get processed in?

File: src/main/resources/swagger/ga4gh-tool-discovery.yaml [hunk 2]
```diff
@@ -381,10 +380,14 @@ definitions:
...
     required:
       - version
+      - api-version
     properties:
       version:
         type: string
         description: The version of this registry
+      api-version:
+        type: string
+        description: The version of the GA4GH supported by this registry
```
Comment (tfmorris) [inline src/main/resources/swagger/ga4gh-tool-discovery.yaml]: Word missing? GA4GH **API** supported by this registry.

File: registry.json (hunk 1 above)
Comment (denis-yuen) [inline registry.json]: If I recall correctly, the order is irrelevant. This is a just a way of pointing our index at a particular server (Dockstore in this case) multiple times if it supports different versions of the API.

**Comment by denis-yuen** [comment]
From the newsgroup:
:+1:  for pagination via offset

File: registry.json (hunk 1 above)
Comment (briandoconnor) [inline registry.json]: It's so a given site (like Dockstore) can advertise multiple versions of the API if need be.  The metadata endpoint includes the API version ID.

**Comment by briandoconnor** [comment]
Yes, pagination sounds like a good idea.  @denis-yuen or others, do you know if anything is built in to swagger to make this easy?  Any particular format we should use (I see requests on the mailing list for using a page offset number which sounds good).

File: src/main/resources/swagger/ga4gh-tool-discovery.yaml (hunk 2 above)
Comment (briandoconnor) [inline src/main/resources/swagger/ga4gh-tool-discovery.yaml]: Thanks @denis-yuen for fixing this typo!

**Comment by tfmorris** [comment]
`offset` & `limit` are pretty classic for pagination. I have a slight preference for offset to be specified in terms of entries as opposed to pages. That way you don't have to multiply/divide by page size, you can just use the desired offset directly. Default to offset=0, limit=100. Only include the total count in the response if it's easy to compute (ie doesn't require enumerating the entire collection, defeating the purpose of pagination).

**Comment by denis-yuen** [comment]
@briandoconnor as far as swagger is concerned, I don't think there are any special annotations for pagination. I think taking the previous post as an example, you just add two query parameters for `offset` and `limit`, describe them, and that's pretty much it.

**Comment by tetron** [comment]
+1 on @tfmorris's suggestion that offset is in items instead of pages, and clients can use limit to set a preferred page size.

**Comment by osallou** [comment]
Regading base path for API:

basePath: /api/v1
paths:
  /tools/{id}:
  ....

I would consider adding a /api/ga4gh/v1 or something alike instead of generic /api/v1.

Indeed, it may clash with existing APIs in tools/registries. ga4gh related API should not overlap with any existing one. Prefixing API with a specific name would avoid this.

**Comment by briandoconnor** [comment]
OK, I _think_ we responded to all comments with tweaks to the schema.

**Comment by osallou** [comment]
one remark/question
/tools return a list of tools. Each tool has different versions and each version propose:
      descriptor:
        $ref: '#/definitions/ToolDescriptor'
      dockerfile:
        $ref: '#/definitions/ToolDockerfile'

this means that each version of each tool will return the Dockerfile and CWL content. This makes result really heavy in size and can make query quite long on server to retrieve all matching tools then versions of each tool with their content.

Maybe it would be enough to return a boolean to say if a Dockerfile/CWL... is available. Then user can query the ga4gh api to retreive the Dockerfile content.

This means the same when querying 1 tool as both /tools and /tools/{id} return Tool instances. Getting Dockerfile/CWL embedded in result could make sense for a single tool query, but not really for all tools query.

**Comment by briandoconnor** [comment]
@osallou makes a good point.  Paging certainly helps but let's discuss on the Monday call if the descriptor and dockerfile should not be embedded in the output for `/tools` or `/tools/{id}` and, instead, be accessible via the distinct descriptor and dockerfile endpoints.  This would reduce the size (and runtime) of the `/tools` response.  Also, we should discuss the correct behaviour when a user requests a page limit that is beyond what the server can handle.  Is that a 404?  Probably the correct code for this condition is 413 or 416?  See https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

**Comment by geoffjentry** [comment]
I agree, let's chat on Monday.

I've seen 404 mostly for these with 416 probably the 2nd most common.

I like the idea of having the tools response be pretty slim, providing info
on what's available. a client can use that information to get the bits that
they want afterwards.

J

**Comment by osallou** [comment]
+1 for 404

**Comment by kislyuk** [comment]
Regarding pagination, I've faced similar design questions and settled on following GitHub's lead: https://developer.github.com/guides/traversing-with-pagination/ - they use the Link header to abstract away the details of pagination parameters. I would recommend this approach.

**Comment by denis-yuen** [comment]
The GitHub design was the inspiration for the links returned in the header https://github.com/ga4gh/tool-registry-schemas/blob/release/1.0.0/src/main/resources/swagger/ga4gh-tool-discovery.yaml#L122 

That said, I would not be surprised if we missed some subtleties.

**Comment by denis-yuen** [comment]
FYI, we've pushed out an update to staging.dockstore.org that allows you to play around with the proposed pagination behaviour at 
https://staging.dockstore.org:8443/api/ga4gh/v1/tools?offset=2&limit=2 (raw response)
https://staging.dockstore.org:8443/static/swagger-ui/index.html#!/GA4GH/toolsGet (allows you to see the response headers)

**Pull request merged**